federation.go

```diff
@@ -0,0 +1,596 @@
+			RoomId:    c.remoteRoomId,
+			SessionId: c.roomSessionId,
+		},
+	})
+}
+
+func (c *FederationClient) updateEventUsers(users []map[string]interface{}, localSessionId string, remoteSessionId string) {
+	for _, u := range users {
+		key := "sessionId"
+		sid, found := u[key]
+		if !found {
+			key := "sessionid"
+			sid, found = u[key]
+		}
+		if found {
+			if sid, ok := sid.(string); ok && sid == remoteSessionId {
+				u[key] = localSessionId
+				break
+			}
+		}
+	}
+}
+
+func (c *FederationClient) updateSessionRecipient(recipient *MessageClientMessageRecipient, localSessionId string, remoteSessionId string) {
+	if recipient != nil && recipient.Type == RecipientTypeSession && remoteSessionId != "" && recipient.SessionId == remoteSessionId {
+		recipient.SessionId = localSessionId
+	}
+}
+
+func (c *FederationClient) updateSessionSender(sender *MessageServerMessageSender, localSessionId string, remoteSessionId string) {
+	if sender != nil && sender.Type == RecipientTypeSession && remoteSessionId != "" && sender.SessionId == remoteSessionId {
+		sender.SessionId = localSessionId
+	}
+}
+
+func (c *FederationClient) processMessage(msg *ServerMessage) {
+	log.Printf("Proxy server message %+v", msg)
+	localSessionId := c.session.PublicId()
+	var remoteSessionId string
+	if hello := c.hello.Load(); hello != nil {
+		remoteSessionId = hello.SessionId
+	}
+
+	var doClose bool
+	switch msg.Type {
+	case "control":
+		c.updateSessionRecipient(msg.Control.Recipient, localSessionId, remoteSessionId)
+		c.updateSessionSender(msg.Control.Sender, localSessionId, remoteSessionId)
+	case "event":
+		switch msg.Event.Target {
+		case "participants":
+			switch msg.Event.Type {
+			case "update":
+				if c.changeRoomId && msg.Event.Update.RoomId == c.remoteRoomId {
+					msg.Event.Update.RoomId = c.roomId
+				}
+				if remoteSessionId != "" {
+					c.updateEventUsers(msg.Event.Update.Changed, localSessionId, remoteSessionId)
+					c.updateEventUsers(msg.Event.Update.Users, localSessionId, remoteSessionId)
+				}
+			case "flags":
+				if c.changeRoomId && msg.Event.Flags.RoomId == c.remoteRoomId {
+					msg.Event.Flags.RoomId = c.roomId
+				}
+				if remoteSessionId != "" && msg.Event.Flags.SessionId == remoteSessionId {
+					msg.Event.Flags.SessionId = localSessionId
+				}
+			case "message":
+				if c.changeRoomId && msg.Event.Message.RoomId == c.remoteRoomId {
+					msg.Event.Message.RoomId = c.roomId
+				}
+			}
+		case "room":
+			switch msg.Event.Type {
+			case "join":
+				if remoteSessionId != "" {
+					for _, j := range msg.Event.Join {
+						if j.SessionId == remoteSessionId {
+							j.SessionId = localSessionId
+							break
+						}
+					}
+				}
+			case "leave":
+				if remoteSessionId != "" {
+					for idx, j := range msg.Event.Leave {
+						if j == remoteSessionId {
+							msg.Event.Leave[idx] = localSessionId
+							if c.closeOnLeave.Load() {
+								doClose = true
+							}
+							break
+						}
+					}
+				}
+			}
```

Comment:
Events with `target: room` can have `type: message` and include a roomid (see, for example, the received signaling message when posting a chat message), but that roomid is currently not converted.

Reply:
Thanks, changed.